bundles/org.openhab.binding.lgthinq/src/main/java/org/openhab/binding/lgthinq/internal/api/TokenManager.java

```diff
@@ -0,0 +1,168 @@
+public class TokenManager {
+    private static final int EXPIRICY_TOLERANCE_SEC = 60;
+    private static final Logger logger = LoggerFactory.getLogger(TokenManager.class);
+    private final OauthLgEmpAuthenticator oAuthAuthenticator;
+    private final ObjectMapper objectMapper = new ObjectMapper();
+    private final Map<String, TokenResult> tokenCached = new ConcurrentHashMap<>();
+
+    public TokenManager(HttpClient httpClient) {
+        oAuthAuthenticator = new OauthLgEmpAuthenticator(httpClient);
+    }
+
+    public boolean isTokenExpired(TokenResult token) {
+        Calendar c = Calendar.getInstance();
+        c.setTime(token.getGeneratedTime());
+        c.add(Calendar.SECOND, token.getExpiresIn() - EXPIRICY_TOLERANCE_SEC);
+        Date expiricyDate = c.getTime();
+        return expiricyDate.before(new Date());
+    }
+
+    @NonNull
+    public TokenResult refreshToken(String bridgeName, TokenResult currentToken) throws RefreshTokenException {
+        try {
+            TokenResult token = oAuthAuthenticator.doRefreshToken(currentToken);
+            objectMapper.writeValue(new File(getConfigDataFileName(bridgeName)), token);
+            return token;
+        } catch (IOException e) {
+            throw new RefreshTokenException("Error refreshing LGThinq token", e);
+        }
+    }
+
+    private String getConfigDataFileName(String bridgeName) {
+        return String.format(THINQ_CONNECTION_DATA_FILE, bridgeName);
+    }
+
+    public boolean isOauthTokenRegistered(String bridgeName) {
+        File tokenFile = new File(getConfigDataFileName(bridgeName));
+        // TODO - check if the file content is valid.
+        return tokenFile.isFile();
+    }
+
+    private String getGatewayUrl(String alternativeGtwServer) {
+        return alternativeGtwServer.isBlank() ? LG_API_GATEWAY_URL_V2
+                : (alternativeGtwServer + LG_API_GATEWAY_SERVICE_PATH_V2);
+    }
+
+    public void oauthFirstRegistration(String bridgeName, String language, String country, String username,
+            String password, String alternativeGtwServer)
+            throws LGThinqGatewayException, PreLoginException, AccountLoginException, TokenException, IOException {
+        LGThinqGateway gw;
+        OauthLgEmpAuthenticator.PreLoginResult preLogin;
+        OauthLgEmpAuthenticator.LoginAccountResult accountLogin;
+        TokenResult token;
+        UserInfo userInfo;
+        try {
+            gw = oAuthAuthenticator.discoverGatewayConfiguration(getGatewayUrl(alternativeGtwServer), language, country,
+                    alternativeGtwServer);
+        } catch (Exception ex) {
+            throw new LGThinqGatewayException(
+                    "Error trying to discovery the LG Gateway Setting for the region informed", ex);
+        }
+
+        try {
+            preLogin = oAuthAuthenticator.preLoginUser(gw, username, password);
+        } catch (Exception ex) {
+            logger.error("Error pre-login with gateway: {}", gw);
```

Comment:
Don't log to error and throw. It would be better to handle it upstream and log it there if essential.
Only log (to debug or trace) details that are only available within this class. Can you adjust it here and in all model files as i see this as a recurring issue.
```suggestion
```